### DISCUSSION TITLE
How to “clip” a frame?

### DISCUSSION THREAD

**Discussion by khaledhosny** [sile-typesetter/sile#1324] [Q&A] [answered]
I’m typesetting some sample text in a frame, I don’t care much about the text I only want the frame to be filled and anything that does not fit thrown away. Currently I set a frame and position it far away from the page and make it `next` for the real frame, so any lines that don’t fit goes to this invisible frame. This seems super hacky and wasteful since all these invsible lines are typeset anyway.

(My use case, is three columns all with the same text but different font sizes, the text is long enough to fill the frame with smallest font size and I don’t want to manually make sure the text fits exactly into each frame, I specify some long enough text and want to be done with it).

**Comment by Omikhleia** [comment, marked as answer]
Quick and dirty hack:
When a frame does not have a next frame and is not the main content frame (which would trigger a new page), then if it overfills it calls `SILE.typesetter.chuck()` which by defaults outputs everything that could not fit (as an emergency).

You could redefine it _not_ to output the extraneous content.
Quick MVP:
```
\begin[papersize=a5, class=book]{document}
\script[src=packages/lorem]

Have I overfilled the running header frame?

\script{
SILE.typesetNaturally(SILE.getFrame("runningHead"), function()
  SILE.typesetter.chuck = function (self)
    SU.warn("Throwing out extra content")
    -- self:leaveHmode(true)
    -- self:outputLinesToPage(self.state.outputQueue)
    self.state.outputQueue = {}
  end
  SILE.call("lorem", { words = 100 })
end)
}

\end{document}
```

Without the `SILE.typesetter.chuck` redefinition, the extra content would be boxed up (via the leaveHmode call) and output to page:

![image](https://user-images.githubusercontent.com/18075640/155821716-799de4a5-d606-4b75-ae68-8357a48914c7.png)

With the redefinition as above:

![image](https://user-images.githubusercontent.com/18075640/155821697-79d629ad-863d-44ab-b518-84768fa5f9fa.png)

An even better solution, possibly, would be to intercept it earlier via `SILE.typesetter.initNextFrame()` (where the above-mentioned logic, i..e. "a frame does not have a next frame and is not the main content frame" etc. is implemented, and to design something more clever. It is left as an exercise to the reader ;) ... but there you could check the frame id for instance, and decide which frames not to 'chuck'...

**Reply by alerque** [reply]
I think adding a `chuck()` function that abandons the content sounds like the right solution. If speed were a concern and there was potentially a lot more content involved there are ways to intercept the processing earlier, but as I suspect this is for a font specimen sheet where perhaps just a few paragraphs work of extra content is dumped this approach is probably good enough.

**Reply by Omikhleia** [reply]
@khaledhosny is the provided answer accepted? Or do you expect something else?

**Reply by khaledhosny** [reply]
I _think_ I tried it and didn’t work in my larger file, but it is a good answer nevertheless. Thanks!
